### PR TITLE
fix: gitea status never green

### LINKED ISSUE #5775: Gitea: status will never be green
There is a bug in lib/platform/gitea/gitea-helper.ts :

https://github.com/renovatebot/renovate/blob/63b5094915bfa85a9de849950fe0b7c749218ad6/lib/platform/gitea/gitea-helper.spec.ts#L792-L797

says recent status is most important 

But it tested in same order as https://github.com/renovatebot/renovate/blob/63b5094915bfa85a9de849950fe0b7c749218ad6/lib/platform/gitea/gitea-helper.ts#L179-L186

If status ordered as  `pending`, `success`,
`getCombinedCommitStatus` will always returns `pending`. @ppmathis 
https://github.com/renovatebot/renovate/blob/63b5094915bfa85a9de849950fe0b7c749218ad6/lib/platform/gitea/gitea-helper.ts#L510-L513

## Fix

change 
https://github.com/renovatebot/renovate/blob/63b5094915bfa85a9de849950fe0b7c749218ad6/lib/platform/gitea/gitea-helper.ts#L511

to
```js
function filterStatus(data) {
  const ret = {};
  for (const i of data) {
    if (
      ret[i.context] &&
      new Date(ret[i.context].created_at) >= new Date(i.created_at)
    ) {
      continue;
    }
    ret[i.context] = i;
  }
  return Object.values(ret);
}
```
```js
for (const cs of filterStatus(res.body)) {
```

### PR DESCRIPTION
**Pull request by NateScarlet**
Closes #5775

### CODE REVIEW

File: lib/platform/gitea/gitea-helper.ts [hunk 1]
```diff
@@ -496,6 +497,21 @@ export const renovateToGiteaStatusMapping: Record<
   red: 'failure',
 };
 
+function filterStatus(data: CommitStatus[]): CommitStatus[] {
+  const ret = {};
```
Comment (viceice) [inline lib/platform/gitea/gitea-helper.ts]:
```suggestion
  const ret: Record<string, CommitStatus>  = {};
```

**Review by viceice — CHANGES_REQUESTED** [review]
needs test fixes

File: lib/platform/gitea/gitea-helper.spec.ts [hunk 2]
```diff
@@ -773,15 +774,27 @@ describe('platform/gitea/gitea-helper', () => {
...
+        'warning',
+        'pending',
+        'success',
+        'unknown',
       ];
 
       const commitStatuses: ght.CommitStatus[] = [
         { ...mockCommitStatus, status: 'unknown' },
       ];
+      const now = new Date();
```
Comment (viceice) [inline lib/platform/gitea/gitea-helper.spec.ts]: Use fixed timestamps for testing, so you can simply use snapshots or equals. Also easier to reproduce

File: lib/platform/gitea/gitea-helper.spec.ts [hunk 3]
```diff
@@ -766,22 +767,104 @@ describe('platform/gitea/gitea-helper', () => {
...
+        {
+          status: 'failure',
+          created_at: '2020-03-25T05:00:00Z',
+          expected: 'unknown',
+        },
+        {
+          status: 'error',
+          created_at: '2020-03-25T06:00:00Z',
+          expected: 'unknown',
+        },
```
Comment (viceice) [inline lib/platform/gitea/gitea-helper.spec.ts]: Is the order guaranteed? Or should we shuffle?

Comment (NateScarlet) [reply lib/platform/gitea/gitea-helper.spec.ts]: We can shuffle manually to add more case,
but not programatically becuase expected value is based on previous added statuses,

Comment (viceice) [reply lib/platform/gitea/gitea-helper.spec.ts]: i think simply reorder would be enough for this test.  so no more tests required

**Review by JamieMagee — DISMISSED** [review]

**Review by viceice — APPROVED** [review]

**Pull request merged**